app/javascript/components/displays.jsx

```diff
@@ -0,0 +1,214 @@
+import SidePanel from '@components/side-panel';
+import Twitter from '@widgets/twitter';
+import Numbers from '@widgets/numbers';
+import Weather from '@widgets/weather';
+import Traffic from '@widgets/traffic';
+
+const Wrapper = styled.div`
+  outline: none;
+`;
+
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
+
+const prefetchDelay = 1500;
+
+export const WidgetDisplay = ({
+  stateCallback,
+  superState,
+  cityName,
+  data,
+}) => {
+  const lastPrefetchTime = useRef(0);
+  const timerId = useRef(null);
+
+  const { location } = data;
+  const { widgets } = location;
+  const { enabled: enabledWidgets, byIdOrFirst: current } = widgets;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const setNextWidgetId = () => {
+    stateCallback(({ transitionTime, pausedTime }) => ({
+      currentWidgetId: nextWidgetId,
+      transitionTime,
+      pausedTime,
+    }));
+  };
+
+  const setPrevWidgetId = () => {
+    stateCallback(({ transitionTime, pausedTime }) => ({
+      currentWidgetId: prevWidgetId,
+      transitionTime,
+      pausedTime,
+    }));
+  };
+
+  const switchPages = ({ keyCode }) => {
+    const upKey = 38;
+    const downKey = 40;
+    const SPACEBAR = 32;
+    const { pausedTime } = superState;
+    if (keyCode === SPACEBAR) {
+      if (pausedTime) {
+        stateCallback(({ currentWidgetId, transitionTime }) => ({
+          currentWidgetId,
+          pausedTime: null,
+          transitionTime: transitionTime - pausedTime + Date.now(),
+        }));
+      } else {
+        stateCallback(({ currentWidgetId, transitionTime }) => ({
+          currentWidgetId,
+          transitionTime,
+          pausedTime: Date.now(),
+        }));
+      }
+    }
+    if (keyCode === downKey) {
+      setNextWidgetId();
+    }
+    if (keyCode === upKey) {
+      setPrevWidgetId();
+    }
+  };
+
+  const handleClicks = ({ target }) => {
+    const targetId = target.id;
+    if (targetId.startsWith('widget_')) {
+      const widgetId = parseInt(targetId.replace('widget_', ''), 10);
+      stateCallback({
+        currentWidgetId: widgetId,
+      });
+    }
+  };
+
+  useEffect(() => {
+    timerId.current = setInterval(() => {
+      const { transitionTime, pausedTime } = superState;
+      if (!transitionTime || pausedTime) {
+        return;
+      }
+      const prefetchTime = transitionTime - prefetchDelay;
+      const now = Date.now();
+
+      if (transitionTime < now) {
+        setNextWidgetId();
+      } else if (
+        prefetchTime < now &&
+        lastPrefetchTime.current < now - prefetchDelay
+      ) {
+        lastPrefetchTime.current = now;
+        setNextWidgetId();
```

Comment:
Is it lost because lastPrefetchTime is defined in the WidgetDisplay and not the WidgetController (so re renders of WidgetController reset lastPrefetchTime)?

Reply:
Its lost because we stopped calling `client.query(...)`

Reply:
Oh I see, I'll revert that back